torch_object_detection_segmenter.py

```diff
@@ -1,31 +1,212 @@
+                - `precise`: crop the part of the image specified by the crop box with the given ``top`` and ``left``.
+                .. warning:: When `precise` is used, ``top`` and ``left`` must be fed valid value.
+            """
+            import PIL.Image as Image
+            assert isinstance(img, Image.Image), 'img must be a PIL.Image'
+            img_w, img_h = img.size
+            if isinstance(target_size, int):
+                target_h = target_w = target_size
+            elif isinstance(target_size, Tuple) and len(target_size) == 2:
+                target_h, target_w = target_size
+            else:
+                raise ValueError(f'target_size should be an integer or a tuple of two integers: {target_size}')
+            w_beg = left
+            h_beg = top
+            if how == 'center':
+                w_beg = int((img_w - target_w) / 2)
+                h_beg = int((img_h - target_h) / 2)
+            elif how == 'random':
+                w_beg = np.random.randint(0, img_w - target_w + 1)
+                h_beg = np.random.randint(0, img_h - target_h + 1)
+            elif how == 'precise':
+                assert (w_beg is not None and h_beg is not None)
+                assert (0 <= w_beg <= (img_w - target_w)), f'left must be within [0, {img_w - target_w}]: {w_beg}'
+                assert (0 <= h_beg <= (img_h - target_h)), f'top must be within [0, {img_h - target_h}]: {h_beg}'
+            else:
+                raise ValueError(f'unknown input how: {how}')
+            if not isinstance(w_beg, int):
+                raise ValueError(f'left must be int number between 0 and {img_w}: {left}')
+            if not isinstance(h_beg, int):
+                raise ValueError(f'top must be int number between 0 and {img_h}: {top}')
+            w_end = w_beg + target_w
+            h_end = h_beg + target_h
+            img = img.crop((w_beg, h_beg, w_end, h_end))
+            return img, h_beg, w_beg
+
+        def _get_input_data(docs: DocumentArray, parameters: dict):
+            traversal_paths = parameters.get('traversal_paths', self._default_traversal_paths)
+            batch_size = parameters.get('batch_size', self._default_batch_size)
+            assert (len(docs) == 2)
+            # traverse thought all documents which have to be processed
+            flat_docs = docs.traverse_flat(traversal_paths)
+            assert (len(flat_docs) == 2)
+            # filter out documents without images
+            filtered_docs = [doc for doc in flat_docs if doc.blob is not None]
+            assert(len(filtered_docs)==2)
+            return _batch_generator(filtered_docs, batch_size)
+
         if not docs:
             return
-        for doc in docs:
-            assert doc.blob.shape[0] == 256
-            doc.embedding = np.random.rand(self._dim)
\ No newline at end of file
+
+
+        batch = _get_input_data(docs, parameters) # a generator of batches of docs
+        for docs_batch in batch:
```

Comment:
```suggestion
       for docs_batch in _get_input_data(docs, parameters):
```

Reply:
I think like this the naming is less confusing

Reply:
and instead of `_get_input_data` u can call it `_get_batch_generator`?

Reply:
what it really is doing is filter input docs, how about `_get_filtered_docs_batch_generator`

Reply:
filter and batching, it makes sense the name yes